v3_async_wip/v3_async_wip/mqtt_client.py

```diff
@@ -0,0 +1,455 @@
+        """
+        return self._connected
+
+    def set_credentials(self, username, password=None):
+        """
+        Set a username and optionally a password for broker authentication.
+
+        Must be called before .connect() to have any effect.
+
+        :param str username: The username for broker authentication
+        :param str password: The password for broker authentication (Optional)
+        """
+        self._mqtt_client.username_pw_set(username=username, password=password)
+
+    async def connect(self):
+        """
+        Connect to the MQTT broker using details set at instantiation.
+
+        :raises: ConnectionFailedError if there is a failure connecting
+        """
+        # Wait for permission to alter the connection
+        async with self._connection_lock:
+            # NOTE: It's not generally safe to use .is_connected() to determine what to do, since
+            # the value could change at any time. However, it IS safe to do so here. The only way
+            # to become connected is to invoke a Paho .connect() and wait for a success. Due to the
+            # fact that this is the only method that can invoke Paho's .connect(), it does not
+            # return until a response is received, and it is protected by the ConnectionLock,
+            # we can be sure that there can't be overlapping invocations of Paho .connect().
+            # Thus, we know that the state will not be changing on us within this block.
+            if not self.is_connected():
+                # Record the type of operation in progress
+                self._connection_lock.connection_type = OP_TYPE_CONNECT
+
+                # Start listening for connect before performing Paho connect to make sure
+                # we don't miss it. This could possibly happen due to timing issues on the
+                # Paho thread (i.e. CONNACK received after invoking connect, but before our
+                # listening tasks have started)
+                async def wait_for_success():
+                    async with self.connected_cond:
+                        await self.connected_cond.wait()
+
+                async def wait_for_failure():
+                    async with self._connect_failed_cond:
+                        await self._connect_failed_cond.wait()
+
+                success = asyncio.create_task(wait_for_success())
+                failure = asyncio.create_task(wait_for_failure())
+
+                # Start the reconnect daemon (if enabled and not already running)
+                if self._auto_reconnect and not self._reconnect_daemon:
+                    self._reconnect_daemon = asyncio.create_task(self._reconnect_loop())
+
+                # Make sure the tasks are running
+                await asyncio.sleep(0)
+
+                # Paho Connect
+                logger.debug("Attempting connect using port {}".format(self._port))
+                try:
+                    rc = await self._loop.run_in_executor(
```

Comment:
Too bad we can't use `asyncio.to_thread`. The syntax is much cleaner.

Reply:
I could backport it if you feel strongly about it - the source code required is very short.

Reply:
No, I don't think it's worth it. It's not that much cleaner.